### PR TITLE
History filtering #382

### PR DESCRIPTION
**Pull request by D0nPiano**
ToDos:
- [ ] Test

Preview:
screenshot from 2017-04-10 21-29-45

### CODE REVIEW

**Comment by tiltec** [comment]
Cool! Should we add a history tab to the stores as well?

File: client/app/components/group/_history/history.controller.js [hunk 1]
```diff
@@ -1,9 +1,112 @@
...
+  }
+
+  showAllStores(bool){
+    angular.forEach(this.stores, (store) => {
+      store.selected = bool;
+    });
+    this.updateFilteredData();
+  }
+
+  getAllStores(){
```
Comment (tiltec) [inline client/app/components/group/_history/history.controller.js]: You could get rid of this function by using the values from the CurrentStores service.

File: client/app/components/group/_history/history.controller.js [hunk 2]
```diff
@@ -1,9 +1,112 @@
...
     });
   }
+
+  $onInit(){
+    this.getAllStores();
+    this.getAllUsers();
+    this.updateFilteredData();
+  }
+
+  update(){
```
Comment (tiltec) [inline client/app/components/group/_history/history.controller.js]: I would try use the angular digest cycle more, e.g. having a function that gets called in the template `getEntries()` and that function returns filtered history data. Advantage: less explicit data management code.

File: client/app/components/group/_history/history.html [hunk 3]
```diff
@@ -1,5 +1,48 @@
...
+            <md-menu-divider></md-menu-divider>
+            <md-menu-item type="checkbox" ng-click="$ctrl.update()" ng-model="user.selected" ng-repeat="user in $ctrl.users">{{user.display_name}}</md-menu-item>
+          </md-menu-content>
+        </md-menu>
+        <md-menu>
+          <button ng-click="$mdMenu.open()">
+            Type
+          </button>
+          <md-menu-content>
+            <md-menu-item type="checkbox" ng-click="$ctrl.update()" ng-model="$ctrl.types.groups">{{'HISTORY.TYPES.GROUPS' | translate}}</md-menu-item>
```
Comment (tiltec) [inline client/app/components/group/_history/history.html]: It is recommended to use the `translate` directive over filters to reduce the watches.
`<span translate="HISTORY.TYPES.GROUPS"></span>`

**Comment by tiltec** [comment]
I'll work on this now

**Comment by tiltec** [comment]
TODO:
(in addition to tests)

- make user/store list scrollable
- make user/store list searchable (it can get long)

File: client/app/components/group/_history/history.controller.js [hunk 4]
```diff
@@ -49,6 +60,77 @@ class HistoryController {
...
+      if (angular.isUndefined(user._selected)) {
+        user._selected = true;
+      }
+      return user;
+    });
+  }
+
+  showAllStores(bool) {
+    angular.forEach(this.getStores(), (store) => {
+      store._selected = bool;
```
Comment (tiltec) [inline client/app/components/group/_history/history.controller.js]: I don't like this, it changes the values in CurrentStores. But everything else seems to involve much more code (e.g. merging together two lists)

Comment (nicksellen) [inline client/app/components/group/_history/history.controller.js]: I would keep selected state in it's own object, something like:

```js
var selectedStores = {};

// select a store
selectedStores[id] = true;

// unselect a store
delete selectedStores[id]

// check if a store is selected
selectedStores[id]
```

Comment (tiltec) [inline client/app/components/group/_history/history.controller.js]: How does that go together with `ng-model`? E.g. here https://github.com/yunity/foodsaving-frontend/pull/449/files#diff-1491a792156b8840cb4c659ffb0de4ddR14

I'm not a huge fan of using getters/setters there, too much boilerplate.

Comment (nicksellen) [inline client/app/components/group/_history/history.controller.js]:
```html
ng-model="selectedStores[store.id]"
```
Maybe?

Sometimes have to use the `ng-click` on a checkbox instead.

Comment (nicksellen) [inline client/app/components/group/_history/history.controller.js]: There are much worse things than boilerplate too.

Comment (tiltec) [inline client/app/components/group/_history/history.controller.js]: Good idea, will refactor now.
(this feature here is not urgent at all, but I wanted to get it merged before the code around it changes too much.. might need to rething on my priorities though)

**Comment by tiltec** [comment]
Didn't manage to enable scrolling in the user list, but otherwise it's ready for merging.

**Pull request merged**